### PR TITLE
Initial check-in of mavenized version of library

### PR DESCRIPTION
**Pull request by wz2b**
Creating a MAVEN version of this library.  If you're interested, suggest you pull it into a new branch (call it "maven" if you wish).  Most of this effort to date is just moving files around and creating the pom.

Still need to:
  * Fix the unit tests - they aren't really unit tests if they depend on having hardware to talk to.  Those are integration tests and should be dealt with separately
  * Figure out where to put some of the 'demo' code.  It shouldn't be in 'tests' and it ideally wouldn't be in the library's main jar.  Ideally there would be a separate 'demos' artifact.  Need to think about what to do here.
  * If the main project is interested in publishing this to maven central, then change the groupId to their own - currently don't know what that would be.

### CODE REVIEW

**Comment by kasemir** [comment]
wz2b de ai4ic: tnx

Well, I personally hate maven, it's made my life miserable over the last couple years.
But if you prefer pain, and we can keep the source tree in an arrangement that still works without Maven, why not?

The group id should be 'org.epics'.
Can you create a pull request that
 * uses 'org.epics.etherip....' for all the package names
 * uses 'org.epics' as the group ID
 * keeps the Eclipse .classpath etc. files instead of deleting them
?

Then
 * I can merge that
 * I'll update the Eclipse files to use the unfortunate Maven directory layout

.. and we're both happy.

73,
Kay

**Comment by wz2b** [comment]
Yes but just to be clear, my goal is to get it into maven central.  Then,
when somebody wants to use it from either gradle or maven, they just
require that dependency.  If I change the groupId to yours, then that means
I can't be the one to push it into maven central - you would have to.  If
you kind of hate maven you might find that process frustrating.


If you don't want to deal with maven central there's another way: you could
just tell the maven people that I'll be releasing into your groupId.  I'm
trustworthy :) but it's really up to you.  I did that for one other project
and they were OK with it.

Putting the eclipse files back is no problem.  Will eclipse be OK with the
fact that I moved source files into src/main/java and src/etst/java?

**Comment by kasemir** [comment]
No problem on changing the 'src' folder info for Eclipse so we can use src/main/java etc.
As for pushing to maven, I'm interested in keeping my sanity, but org.epics would still be the most suitable group. Will ask the other contributors to org.epics, stay tuned.

**Comment by wz2b** [comment]
Do you by chance know if it's possible to access a tag that's not a
Controller level tag but within a program (within the PLC)?

**Comment by kasemir** [comment]
I think it's limited to controller tags.
This includes those that you create and also the auto-generated ones for the I/O modules.
It does not include program tags.

**Comment by wz2b** [comment]
Oh ok.

Do you know if this library also lets you read custom data types, or
arrays, or arrays of custom data types?

**Comment by kasemir** [comment]
Have you tried it at all, or are you suggesting to first restructure everything without any idea if you need it at all ;-) ?

This is the Java version of a C based implementation that has been in production for about a decade in the form of a device driver for EPICS IOCs.
The C version has a document, https://github.com/EPICSTools/ether_ip/tree/master/ether_ipApp/doc, that gives more details.

It can access elements of structures, like "Local:2:I.Ch0Data" for those automatically created controller tags of I/O modules. Likewise it can access elements of custom structures, one element at a time, as long as those are then basic types.
It can also access array elements, or complete arrays.
Check the documentation for the C version on why we actually prefer arrays, with limitations.
Reading a complete array is about the same speed as reading a single element, so we pack data in arrays and then use alias tags in the ladder logic to make that more readable. But there's a limit to the array size based on a buffer limit in the PLC of abour 500 bytes.

**Comment by wz2b** [comment]
Yeah.  Actually, everything I tried so far worked except:

  - I didn't know you can only read controller-level tags ...
  - I did try to read the array but it returned NULL ... maybe I just don't
know what tag name to use.

Thanks for the reference to that doc - I will read through it before I ask
any more stupid questions.

What I'm doing now is testing the mavenized version.  So far so good!  I
guess that's to be expected since what I'm doing is only packaging and not
touching any of the code (other than integration tests, and I still haven't
figured out what to recommend we do with those).  I made a Spring Boot
application as a test.  It makes a simple REST interface to fetch tags
using the etherip library.  It's a tiny amount of code to be able to:

    GET http://localhost/192.168.1.15/0/MotorStatus

and have it reach out and fetch that.  To make for a more complete
demonstration, though, I want to make sure it can handle (and serialize to
json/xml/yaml) any of the types that get returned, even if it's something
more complex than just a DINT or BOOL.

This will be something I can push to github as a demo when I'm done...
since I"m working on it on my own time (not for work).

Sorry to be a pest, I'll try to be quiet :)

--Chris

**Comment by kasemir** [comment]
Let me rephrase the access to arrays and structures:
The protocol supports it, the C version supports it. The java version for now doesn't allow you to say how many array elements you want (see #6) and there's a quirk with parsing the "." in structure tags (#8) which should both get fixed.

As for the controller vs. program tags, it's based on the documented protocol elements. Using those, I could only see controller tags. Maybe there's a way to access program tags with some magic syntax for the name, but that's not documented. Likewise I found no hint on how to list tags. The proprietary AllenBradley/Rockwell software can list tags and monitor program tags, but that's not described in "Logix5000 Data Access".

Your REST interface sounds like a great idea.
Another one would be an MQTT connection, something that periodically publishes the values read from a PLC to MQTT, which makes it accessible to more IoT clients.

**Comment by wz2b** [comment]
Sure.  i could do an MQTT one... reactivex would be fun too.  Let me see
how far I get.

My array is called DataTable1 and it is 100 elements long.  Each element
has:
   T : DINT
   X:  DINT

when I try to get DataTable1, it throws:

java.lang.Exception: Unknown CIP type code 0xfffff154
at etherip.types.CIPData$Type.forCode(CIPData.java:55)

(really iblowing up in
CIPReadDataProtocol.decode(CIPReadDataProtocol.java:50))


When I try to get just ONE element, I just get al og entry:
        Read Tag 'DataTable1[0]' (no value)

so when I asked the original question I was really just trying to figure
out what's going on.  It's a 5370 controller,

#8 does look related.  I need to figure out how to turn finest logigng on -
the logigng configuration in spring boot is a weird combination of slf4j
log4j and others.  Let me do that now.

**Comment by wz2b** [comment]
Hi,

In the session, where does "Funstuff" come from?  I can't find that word in
the etherip source, so I'm guessing it must come from the PLC itself.

After reading through the ether_ipApp/doc info and both those tickets I
decided to try A simpler test.  I created:

    intarr : DINT[8]


I think (from docs) both of these things should work:

   tags.add("intarr");
   tags.add("intarr[3]");


but I get results Id on't understand.

final String tagName = "intarr";

final TagList tags = new TagList();

tags.add(tagName);

tags.process(connection);

CIPData result = tags.get(tagName).getValue();

int elementCount = result.getElementCount();

Number result2 = result.getNumber(2);


elementCount comes back as 1, when I think it should be 8.

When I try to get "intarr[0]" (or any other subscript) I just get nothing
back as if it doesn't exist:

   etherip.EtherNetIP          : Read Tag 'intarr[1]' (no value)


Could it be that I somehow created the tag in the PLC incorrectly?

If you want me to open a ticket I can do that but I have the feeling
there's nothing wrong with the software, I'm just doing something stupid.

**Comment by kasemir** [comment]
Funstuff is what I put into the encapsulation header as a 'context', see
https://github.com/EPICSTools/etherip/blob/master/src/etherip/protocol/Encapsulation.java#L75

For the arrays, you need to tell it how many elements you want to read.
Right now it always just asks for one element.
See ticket 6, `CIPReadDataProtocol.encode()` you'll find
````
        buf.putShort((short) 1); // elements
````
As mentioned in that ticket, we need to pass the elements down so you could ask for tag "intarr" and 8 elements. I don't think you can ask for "all" elements, you need to know the tag name as well as 
the size and then ask for it.

Access to one element as tag "intarr[1]" should work, that could be part of ticket 8 where the tag name is not properly parsed. That works in the C version, but as you mentioned this Java version is a side project of mine that's not really used, so I don't get around to fixing this.

**Comment by wz2b** [comment]
I can definitely fix that.  I have time to do it now.  I'm just looking at
the stack trace from EthernerNetIP.readTag() down to
CIPReadDataProtocol.encode() to figure out how many layers I need to
actually touch.

**Comment by wz2b** [comment]
OK, here's a description of what I did.

EtherIP:  added

**Comment by wz2b** [comment]
sorry, I don't know how Is just sent you a mostly empty e-mail.  This seems
to all work fine .. but I can't send it to you as a pull request until I
undo a few things including changing all the packages to org.epics as we
had previously discussed.  I don't know what I'm going to do to get you
back to a pull request that doesn't throw your whole codebase into chaos
... but I'll figure something out.  Meanwhile here (below) is what I did.


I still can't read my custom struct.  The PLC calls it a 'user defined
datatype.'  When I try to read it, regardless of array or not, the library
still gives me an exception:

    java.lang.Exception: Unknown CIP type code 0xfffff154

and I don't know what that means.  Does the library somehow have to get a
type definition first before trying to retrieve the value?  There's no
example of that on the java side... I"ll go look and see if there's one for
the C library.

Meanwhile, here's what I did to fix bug 6 - if I did something stupid
please let me know.






EtherIP: added

public CIPData readTag(final String tag, int count) throws Exception

{
   final MRChipReadProtocol cip_read = new MRChipReadProtocol(tag, count);
   final Encapsulation encap =
         new Encapsulation(SendRRData, connection.getSession(),
               new SendRRDataProtocol(
                     new UnconnectedSendProtocol(slot,
                           cip_read)));
   connection.execute(encap);

   return cip_read.getData();
}


MRChipReadProtocol  added a constructor:


public MRChipReadProtocol(final String tag, final int count)
{
    this(tag, new CIPReadDataProtocol(count));
}


CIPReadDataProtocol  added a cosntructor and a field:

private final int count;

public CIPReadDataProtocol(int count){
    this.count = count;
}


Then modified encode to use this field when encoding the message:

    buf.putShort((short) count); // elements

**Comment by wz2b** [comment]
OK.  I'm getting further.

I modifeid the CIPData enum to create a real reverse map for lookups:

final private static Map<Short, Type> reverse;

static {
    reverse = new HashMap<>();
    for (Type t : EnumSet.allOf(Type.class)) {
        reverse.put(t.code, t);
    }
}

public static Type forCode(final short code) throws Exception {
    Type t = reverse.get(code);
    if (reverse == null) {
        throw new Exception("Unknown CIP type code 0x" +
Integer.toHexString(code));
    }
    return t;

}




By doing that I got this:

   java.lang.Exception: Structure elements of type STRUCT (0x02A0) not
handled

AH hah!  Chasing down that code:

case STRUCT: {
    final Type el_type = Type.forCode(data.getShort(0));
    if (el_type == Type.STRUCT_STRING)
        return 1;
    else
        throw new Exception("Structure elements of type " + type + "
not handled");



So it seems like it can only handle structs of ... only strings?  or ..
it's weird because the error makes it seem like I have a struct of structs,
which I don't....

**Comment by kasemir** [comment]
I never tried to handle structs, because their content was unclear.
I do support reading and writing strings, which turned out to be one type of struct.

**Comment by wz2b** [comment]
thanks.  Maybe I'll look at how to handlle structs but I have a feeling it
may be more than I want to get into.  I'm only going to have access to an
A-B PLC to test with until mid-JAN.

**Comment by kasemir** [comment]
Reading structs has not been a requirement for us as long as reading "some_struct.element" is supported - which it is in the C version, and should work in the Java version except for issue 8.
So fixing that would meet most needs, just fetch elements one by one, instead of trying to decode them on the client side, which may be quite hard, at least I didn't find that documented.

**Comment by wz2b** [comment]
So if the struct instance is called dptest and it has attributes T, and X
then I should be able to:

    CIPData intarr = plc.readTag("dptest.T");


and that should work?  It seems to always return null.  I'm trying to trace
through why.

**Comment by kasemir** [comment]
Right, "dptest.T" should read element T of structure tag dptest and succeed as long as T is a basic type like REAL, DINT, ..
That works with the C version, and issue 8 indicates that something was missing with parding the tag into path elements.

**Comment by wz2b** [comment]
OK.  I figured out why.

WHen I request dptest.T it sends this:

0000 - 6F 00 2C 00 01 00 54 00 00 00 00 00 46 75 6E 73 - o.,...T.....Funs
0010 - 74 75 66 66 00 00 00 00 00 00 00 00 00 00 02 00 - tuff............
0020 - 00 00 00 00 B2 00 1C 00 52 02 20 06 24 01 0A F0 - ........R. .$...
0030 - 0E 00 4C 05 **91 08 64 70 74 65 73 74 2E 54** 01 00 - ..L...dptest.T..
0040 - 01 00 01 00                                     - ....   


Note that it sent "dptest.T" as one type 91 (which I think means tag description).

According to [This Document](http://literature.rockwellautomation.com/idc/groups/literature/documents/pm/1756-pm020_-en-p.pdf) page 64.  According to this, what the request should contain is this:
`
  91 06 64 70 74 65 73 00          /* tag, 6 bytes, dptest */
  91 01 54                /* tag, 1 byte, T */
`
I think what has to happen is to split on the period then add a null (?) then another 91 and length ... I am looking for the documentation that explains what that 00 (I interpret it as a null but it's really not) means.  The document I quoted above shows examples but I can't see where it explains that part.

**Comment by kasemir** [comment]
Yes. Read Issue #8. It explains how the CNPath needs to change into a CNSymbolMultiPath that splits a tag name at ".".
In the C version, https://github.com/EPICSTools/ether_ip/blob/master/ether_ipApp/src/ether_ip.c#L362 , note how `EIP_parse_tag` splits the tags at "." for structure elements or "[" for array elements.

**Comment by wz2b** [comment]
Just a correction to what I typed earlier: That 00 is actually a pad byte.  The documentaiton says:

> (1) Optional pad byte (00) when length of string (length) is odd. (The pad byte is not added to the length.)

so my example was wrong.  My string is even length so no pad there.  So for dptest.T it should be:

91 06 64 70 74 65 73      (dptest)
91 01 54 00                      (T)

**Comment by kasemir** [comment]
Alright, are you done with this one?
Then I'll merge & update the Eclipse files.

**Comment by wz2b** [comment]
Not yet.  I am going to make two more pull requests:

- One will leave your code base "as is" but add the change to the 'dot' and
'bracket' paths so that it works right with structure-members and
single-element-array access

- The second pull request, which I'll make afterwards, will be the
mavenized version but with everything put into org.epics AND with your
eclipse settings put back in.  That will be a different tag/revision.  I'm
not sure if that means I have to make a second pull request at that
revision or not, but either way I'm not ready yet.

**Comment by kasemir** [comment]
Thank you, this is good news! I'd actually prefer a separate PR for the dot/bracket tag parsing, and starting over with the mavenization by keeping the existing .* files for Eclipse is also better.
We can keep this PR as a discussion/log, but otherwise we won't merge this one because it'll be replaced by what you just outlined.

**Comment by wz2b** [comment]
This pull request will be resubmitted at a later date so that it rolls in other library changes made in #11
